keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,276 @@
+
+
+```golang
+	var label_value string
+	for i=0;i<1000000;i++ {
+		label_value = strconv.Itoa(i+4) // bring us out of our expected range
+		SomeMetric.WithLabelValues(label_value).Inc()
+	}
+```
+
+Then in existing Kubernetes components, we would have a terrible memory leak, since we would effectively create a million metrics (one per unique label value). If we curled our prometheus endpoint it would thus look something like this:
+
+```prometheus
+# HELP some_metric alalalala
+# TYPE some_metric counter
+some_metric{label_too_many_values="1"} 1
+some_metric{label_too_many_values="2"} 1
+some_metric{label_too_many_values="3"} 1
+some_metric{label_too_many_values="4"} 1
+... //
+... // zillion metrics here
+some_metric{label_too_many_values="1000003"} 1
+```
+
+That would suck. With our cardinality enforcer in place, we would expect the output to look like this:
+
+```prometheus
+# HELP some_metric alalalala
+# TYPE some_metric counter
+some_metric{label_too_many_values="1"} 1
+some_metric{label_too_many_values="2"} 1
+some_metric{label_too_many_values="3"} 1
+some_metric{label_too_many_values="unexpected"} 1000000
+```
+
+We can effect this change by registering a whitelist to our prom registries during the boot-up sequence. Since we intercept metric registration events and have a wrapper around each of the primitive prometheus metric types, we can effectively add some code that looks like this (disclaimer: the below is pseudocode for demonstration):
+
+```golang
+const (
+	unexpectedLabel = "unexpected"
+)
+
+func (v *CounterVec) WithLabelValues(lvs ...string) CounterMetric {
+	if !v.IsCreated() {
+		return noop // return no-op counter
+	}
+	// INJECTED CODE HERE
+	newLabels = make([]string, len(lvs))
+	for i, l := range lvs {
+
+		// do we have a whitelist on any of the labels for this metric?
+		if metricLabelWhitelist, ok := MetricsLabelWhitelist[BuildFQName(v.CounterOpts.Namespace, v.CounterOpts.Subsystem, v.CounterOpts.Name)]; ok {
+			// do we have a whitelist for this label on this metric?
+			if whitelist, ok := metricLabelWhitelist[v.originalLabels[i]]; ok {
+				if whitelist.Has(l) {
+					newLabels[i] = l
+					continue
+				}
+			}
+		}
+		newLabels[i] = unexpectedLabel
+	}
+
+	return v.CounterVec.WithLabelValues(newLabels...)
+	// END INJECTED CODE
+}
+```
+
+This design allows us to optionally adopt @lilic's excellent idea about simplifying the interface for component owners, who can then opt to just specify a metric and label pair *without* having to specify a whitelist. Personally, I like that idea since it simplifies how a component owner can implement our cardinality enforcing helpers without having to necessary plumb through complicated maps. This would make it considerably easier to feed this data in through the command line since you could do something like this:
+
+```bash
+$ kube-apiserver --bind-metric-labels "some_metric=label_too_many_values"
```

Comment:
I'm not terribly crazy about my wording on the flag here.

Reply:
How about `--supported-label-values` or `--accepted-label-values`?